pkg/comments/builder.go

```diff
@@ -0,0 +1,61 @@
+package comments
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/crossplane-contrib/terrajet/pkg/markers"
+)
+
+const (
+	errCannotParseAsFieldMarker = "cannot parse line as marker for field"
+)
+
+// Builder builds comments for fields by also building terrajet options
+// supported by comment markers
+type Builder struct {
```

Comment:
It seems that we're utilizing `controller-tools` marker parsing mechanisms but do you think we need this tool to be that generic given that it lives in `terrajet` repo, not `crossplane-tools`? What do you think about having fields for each possible marker `terrajet` knows, like `Required bool`, and use that as medium for `Builder` here? Very similar to current design but instead of only `CRDTags`, there would be all other options we know (required, minimum, reference etc.) as fields in `Options` and developers would just need to worry about this single `Builder` struct that has an `Options` field. The `Builder` would only parse/build the options it knows of and unrecognized lines would be treated just like any other comment text.

It's a rough general idea, I'm just wondering if we can leverage the fact that we're in a specific repo to make the operation more strong-typed and easy to work with since we don't have to pay the cost of being so generic like `crossplane-tools` or `controller-tools`.

Reply:
Actually, I am not trying to be generic here, rather trying to keep things scoped. This `comments builder` is responsible for building added comment lines as proper comments, e.g. Regular comments listed before marker ones. One exception here, which I am also not much happy about is, it can also build terrajet marker options if input comments have markers that terrajet recognizes. The `markers` package on the other hand is responsible for handling terrajet recognized markers and means to build kubebuilder markers with strongly typed structs.

Regarding kubebuilder options, I also considered having possible configurations in a single struct, but after checking supported kubebuilder markers (e.g. for [validation](https://book.kubebuilder.io/reference/markers/crd-validation.html)), I didn't prefer a struct growing with all possible parameters there. I believe the current approach is straightforward enough with the helpers added (e.g. `markers.MarkerForConfig()`) but still extendible especially considering there could be some markers that require (multiple) parameters. 

If we really want a struct with support options, I would rather prefer to build another layer on top which uses the above mechanisms, but I am hesitant that it could grow very quickly with a lot of options.

Reply:
> Regarding kubebuilder options, I also considered having possible configurations in a single struct, but after checking supported kubebuilder markers (e.g. for validation), I didn't prefer a struct growing with all possible parameters there.


If we want to cover all kubebuilder markers, yes, but I'm not sure that's where we'll end up. The use cases I'm able to come up for this comment utility:
* Let users add markers Terrajet recognizes, like overriding a field tag.
* Let Terrajet read markers it can act upon, like overriding a field tag. 
* Let Terrajet add markers for certain validations derived from terraform.Schema, like kubebuilder `required` marker.

In all cases, we actually have enough information to scope the options. For example, the set of `kubebuilder` tags we'll ever need is limited by what [TF Schema](https://github.com/hashicorp/terraform-plugin-sdk/blob/9321fe1/helper/schema/schema.go#L37) exposes, which seems like `required`, min/max, and possibly input default. The rest isn't really needed to be recognized by Terrajet because it won't take any action anyway. For other two use cases, Terrajet is either writer or reader of the same information.

If we have all this in a strong-typed medium, we can use that type both as internal representation of a comment in Terrajet and also exception input for specific fields from the author of provider.

Reply:
@muvaf refactored for a simpler interface considering the limited requirements we have here, as we also discussed offline. I dropped support for multiple keys in a single marker to simplify further (completely removing dependency to controller-tools etc.), and we can consider adding support for that when/if we need that functionality.

Reply:
@turkenh Thank you!